blogposts/2020/code-search-turned-code-checker.md

```diff
@@ -0,0 +1,296 @@
+
+Using the test files isn't a guarantee that I've implemented all the checks or
+that it's entirely precise, but it adds a lot more confidence than inferring
+patterns only from documentation.
+
+### Results
+
+Without making any claims about completeness, I was able to implement
+at least one variant for 19 out of 34 checks checks that I feel confident
+about. I relied only on patterns in `staticcheck`'s test data to discover syntactic
+variants for checks, so I don't know if I covered all the patterns that
+`staticcheck` implements in its code.
+
+The majority of checks that I couldn't write required type information to
+implement correctly (13 of 34). Other checks I couldn't write required more
+complex syntax matching rules (8 of 34). This table roughly quantifies the
+expressive needs for implementing checks:
+
+| Description                      | # of checks |
+|----------------------------------|-------------|
+| Total                            | 34          |
+| Works (all variants)             | 11          |
+| Works (at least one variant)     | 19          |
+| Needs additional syntax matching | 8           |
+| Needs type info                  | 13          |
+
+<sup>Note that some checks require type info _and_ additional syntax matching.
+Also, one working variant for a check may not require type info, but another
+variant for the same check may. I.e., the values overlap and do not sum to
+Total.</sup>
+
+Extending search queries to access static properties like type information is a
+natural extension for writing better code checks, and an area of [code intelligence](https://docs.sourcegraph.com/user/code_intelligence)
+work that we're exploring.
+
+It was tempting to compare more directly with `staticcheck` output by
+downloading the 100 Go repositories to disk, and running `staticcheck` on them
+to see how the same patterns are found. For `staticcheck` to be effective, the
+project typically needs to be built first (my experience was that running
+`staticcheck` on individual files can be hit-and-miss, and understandably so). I
+didn't like the idea of doing all that work, so I punted. It does nicely motivate
+why less precise, but quicker code check workflows appeal.<sup>5</sup>
+
+### Takeaway
+
+Code search can be much more than finding your favorite provider called
```

Comment:
Maybe a table showing relative advantages/disadvantages of staticcheck vs seach+comby would be evocative here.

Potential rows:
* Accuracy (syntax vs semantics)
* Scalability
* Ease of definition
* Ease of execution

Reply:
I like this idea in principle (thanks for suggesting it), but I feel like all of these bullets needs to be better qualified to make a valid point, and anything short of that becomes handwavy. Example:

Ease of definition: this depends a lot on who writes the definition, and who uses it. A user of staticcheck doesn't really care how easy/hard it is to define checks in that framework. Sure it's nice that they can define their own with search, but they may not be able to think of anything off the top of their mind (much nicer if someone just does the thinking for you and you push a button), so the perceived benefit to a casual user is a bit moot unless we dig into it.

If we do make have a qualified discussion, I think it detracts from a simpler takeaway ("here's an example, see this can work instead?") that applies to the audience broadly, not just people who are familiar with scalability